sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2661,7 +2657,12 @@ class Dataset[T] private[sql](
    */
   def toLocalIterator(): java.util.Iterator[T] = {
     withAction("toLocalIterator", queryExecution) { plan =>
-      plan.executeToIterator().map(boundEnc.fromRow).asJava
+      val objProj = GenerateSafeProjection.generate(deserializer :: Nil)
```

Comment:
It should be better to explain we keep the projection inside for thread-safe.

Reply:
+1